### PR TITLE
Move CIs to use docker builds

### PR DESCRIPTION
**Pull request by powderluv**
Now that #1234 has landed and anyone can run CI / Release builds locally move GHA to use the same flow.

### CODE REVIEW

**Review by silvasean — CHANGES_REQUESTED** [review]

File: .github/workflows/buildAndTest.yml
Comment (powderluv) [reply .github/workflows/buildAndTest.yml:164]: yeah I have https://github.com/llvm/torch-mlir/issues/1317 to do it only for generated files.

Comment (silvasean) [reply .github/workflows/buildAndTest.yml:164]: I don't think filing an issue is sufficient. Can you please address it in this PR so that we don't regress?

Comment (powderluv) [reply .github/workflows/buildAndTest.yml:164]: I have some questions on what is the expected behaviour of the test. Will post in the issue.

Comment (powderluv) [reply .github/workflows/buildAndTest.yml:164]: We didn't have that check less than 24hrs ago and it is a blind check `git diff --quiet`. I wish we waited for my review on https://github.com/llvm/torch-mlir/pull/1286 less than 24hrs ago. Now I need to take the burden of that commit along with the docker build. 
Let me know if it is acceptable to address #1317 separately and make progress on this. I can take on #1317 next.

Comment (silvasean) [reply .github/workflows/buildAndTest.yml:164]: I think we should just copy the current check into the script. For now, the script isn't really useful for local iteration since it runs unit tests and all e2e tests on all backends, which is not really a useful configuration for incremental development (way too slow, and doesn't support test filtering/etc. and most incremental changes only need testing on one backend). We will need to refine the script for use in incremental development as a second step.

Comment (powderluv) [reply .github/workflows/buildAndTest.yml:164]: We spent a lot of time to ensure iterative builds work well and fast. Please see the docker commit message with all the runs and their times. 

Copying this test to break iterative builds (and the hundreds of test runs we did) is discouraging. 

I offered a pragmatic solution to fix the test separately and filed the issue. (even though it landed yesterday without my feedback as someone who spent more time on the CI than I would like). 

So if you insist let me know we will add the test first and delay the docker switch over. 

Please get my feedback or @sjain-stanford on CI commits meanwhile (at least until someone takes over CI maintenance). It adds a higher burden with an already high one to keep all users happy.

Comment (sjain-stanford) [reply .github/workflows/buildAndTest.yml:164]: > Yep, it should be easy to list the specific files. It is currently ShapeLibrary.cpp and GeneratedTorchOps.td

@powderluv If it's not too difficult, can we specify these files so it's not a blind `git diff` check? This way we can satisfy both constraints (avoid false positives from local unstaged changes, and not regress this check from GHA CI).

> the script isn't really useful for local iteration since it runs unit tests and all e2e tests on all backends, which is not really a useful configuration for incremental development (way too slow, and doesn't support test filtering/etc. and most incremental changes only need testing on one backend). We will need to refine the script for use in incremental development as a second step

I do agree with this pain point. To enable incremental development, we may have to spawn a container in interactive mode (bash entrypoint), so folks can do all sorts of incremental edit / compile / edit iterations and run tests selectively within the container. Of course this is follow-on work.

Again, very appreciate of your efforts @powderluv, and thanks @silvasean for the valuable feedback.

Comment (silvasean) [reply .github/workflows/buildAndTest.yml:164]: First of all, I want to thank you @powderluv for all the work on this, and it's unfortunate that we had a race here.

I think we should distinguish "incremental builds" and "incremental development" here.

"incremental builds" means that the creation of build artifacts is accelerated based on a previous build

"incremental development" means a workflow suitable for developers to use as part of a tight edit/compile/test cycle. This may involve a dirty work tree or other weird in-flight situations, like a local hack to a generated file. This also involves running filtered tests etc. The set of tests that make sense to run for CI do not make sense to run for "incremental development".

We can add include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td and lib/Dialect/Torch/Transforms/ShapeLibrary.cpp to the `git diff` command line, but that does not actually fundamentally fix anything, which is that this check only makes sense as part of CI. The invariant that is checked only makes sense to check for a "prospective ready-to-be-merged repo state". For example, I might be iterating locally and make a hacky change to a generated file to see if that fixes an issue I'm having.

I think I have a middle-ground fix that somewhat helps the "run the CI checks, but with a dirty work tree", so let's land this so I can do that. It's pointless for us to be writing walls of text here over what is 10 minutes of coding either way.

Comment (powderluv) [reply .github/workflows/buildAndTest.yml:164]: ok so I have https://github.com/llvm/torch-mlir/issues/1319 for capturing what we can / can't do in "incremental development". Currently I use the docker build for exactly that but maybe I am not a power user enough with hand edited generated files etc. So based on that feedback I can help drive those features. 

I also have https://github.com/llvm/torch-mlir/pull/1318 to fix Issue 1317 that I filed for moving the shapelib / torch-ods.  It has the benefit of not being tied to the CI but will tell the end user whenever they run it if there is any change.  Once that lands we can revisit this PR.

Comment (powderluv) [reply .github/workflows/buildAndTest.yml:164]: #1318 has landed. So will resolve this thread.

**Review by silvasean — APPROVED** [review]

File: .github/workflows/buildAndTest.yml
Comment (sjain-stanford) [inline .github/workflows/buildAndTest.yml:60]: nit: since this is no longer just a "configuration" but includes build + tests, could we rename this to an appropriate name?

Comment (sjain-stanford) [inline .github/workflows/buildAndTest.yml:60]: Also, `llvm-build='in-tree'` should probably change to `llvm-build='${{ matrix.llvm-build }}'`

**Review by sjain-stanford — APPROVED** [review]
LGTM with some nit requests.

**Comment by powderluv** [comment]
@silvasean if you have a quick fix for https://github.com/llvm/torch-mlir/issues/1334 lets land that and then this change. If not we can revert the last shape lib test and re-land after this lands. 

I would like to try and land this today so I can tend to the CI over the long weekend in case anything needs tweaking / reverts etc.

**Comment by silvasean** [comment]
Let's land this! The fix is in https://github.com/llvm/torch-mlir/pull/1338 -- feel free to merge it yourself if needed.

**Comment by powderluv** [comment]
awesome thanks. will wait for it to go green and land #1338 and then land this when green.

**Comment by silvasean** [comment]
Awesome. Thanks Anush for all the hard work here and sorry for these last minute speed bumps on the generated files!

**Pull request merged**
